core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -49,14 +49,211 @@
 #ifdef KOKKOS_ENABLE_CUDA
 #include <cuda_fp16.h>
 #include <cstdint>
+#include <cstdio>
 
 #ifndef KOKKOS_IMPL_HALF_TYPE_DEFINED
 // Make sure no one else tries to define half_t
 #define KOKKOS_IMPL_HALF_TYPE_DEFINED
 
 namespace Kokkos {
 namespace Experimental {
-using half_t                       = __half;
+
+// Forward declarations
+class half_t;
+
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(float val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(double val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(short val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(int val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(long val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(long long val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned short val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned int val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned long val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned long long val);
+
+class half_t {
+  __half val;
+
+public:
+  KOKKOS_FORCEINLINE_FUNCTION
+  half_t():val(__float2half(0.f)){}
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  operator __half() const {
+    return val;
+  }
+
+// Changing below to 1 produces the constructor overload error
+#if 0
+  KOKKOS_FORCEINLINE_FUNCTION
+  operator bool() const {
+    return __half2float(val) == 0.0F;
+  }
+#endif
+
+  KOKKOS_INLINE_FUNCTION
+  half_t(__half rhs):val(rhs) {}
+
+#if 1
+  // Cast rhs to half for assignment to lhs of type half_t
+  template <class T>
+  KOKKOS_INLINE_FUNCTION
+  half_t(T rhs):val(cast_to_half(rhs)) {}
+#endif
+
+  // Unary operators
+  KOKKOS_INLINE_FUNCTION
+  half_t operator+ () const {
+    half_t tmp = *this;
+    #ifdef __CUDA_ARCH__
+      printf("half_t unary operator+\n");
+      tmp.val = +tmp.val;
+    #else
+      printf("float unary operator+\n");
+      tmp.val = __float2half(+__half2float(tmp.val));
+    #endif
+    return tmp;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  half_t operator- () const {
+    half_t tmp = *this;
+    #ifdef __CUDA_ARCH__
+      printf("half_t unary operator-\n");
+      tmp.val = -tmp.val;
+    #else
+      printf("float unary operator-\n");
+      tmp.val = __float2half(-__half2float(tmp.val));
+    #endif
+    return tmp;
+  }
+
+  // Binary operators
+  KOKKOS_INLINE_FUNCTION
+  half_t operator= (__half rhs) {
```

Comment:
Shouldn't it be returning a reference?

Reply:
I don't think so since this is a scalar type. Thoughts?

Reply:
This needs to return a reference. Consider this code:
```
#include<cstdio>
#include<cstdlib>
int main(int argc, char* argv[]) {
  double x = 1.0, y=2.0;
  double* ptr = &(x=y);
  printf("%p %p %lf\n",ptr,&x,x);
}
```
Prints:
```
0x7fffe4133b18 0x7fffe4133b18 2.000000
```